<a href="https://colab.research.google.com/github/DaudAhmad0303/LeetCode-Problems-Solution-colab/blob/master/LeetCode_Problems_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Random Practise Codes**

In [ ]:
nums = [4,1,2,3,8]
print(nums[::2])
print(nums[::-2])
print(nums[::-1])

[4, 2, 8]
[8, 2, 4]
[8, 3, 2, 1, 4]


# **Random Questions Solved**

## **2164. Sort Even and Odd Indices Independently**

In [ ]:
class Solution:
    def sortEvenOdd(self, nums):
        if len(nums) <= 2:
            return nums
        odd = []
        even = []
        l = len(nums)
        for i in range(len(nums)):
            if i % 2 == 0:
                even.append(nums[i])
            else:
                odd.append(nums[i])
        odd.sort(reverse = True)        
        even.sort()
        nums = []
        for i in range(len(odd)):
            nums.append(even[i])
            nums.append(odd[i])
        if l % 2 != 0:
            nums.append(even[-1])
        return nums
obj = Solution()
print(obj.sortEvenOdd(nums = [4,1,2,3]))

[2, 3, 4, 1]


In [ ]:
class Solution:
    def sortEvenOdd(self, nums):
        odds = nums[::2]
        evens = nums[1::2]
        odds.sort()
        evens.sort()
        evens.reverse()
        odds.append(-1)
        evens.append(-1)
        
        res = []
        for a,b in zip(odds, evens):
            if a >= 0:
                res.append(a)
            if b >= 0:
                res.append(b)
        
        return res
obj = Solution()
print(obj.sortEvenOdd(nums = [4,1,2,3]))

[2, 3, 4, 1]


## **125. Valid Palindrome**

In [ ]:
class Solution:
    def isPalindrome(self, s: str) -> bool:
        st = ""
        for i in range(len(s)):
            if s[i].isalpha() or s[i].isdigit():
                st += s[i].lower()
        print(st)
        return st == st[::-1]
obj = Solution()
print(obj.isPalindrome(s = "A man, a plan, a canal: Panama"))

amanaplanacanalpanama
True


## **Count Operations to Obtain Zero**

In [ ]:
class Solution:
    def countOperations(self, num1: int, num2: int) -> int:
        if num1 != num2:
            return min(num1, num2) + abs(num1-num2)
        elif num1 == 0 or num2 == 0:
            return 0
        else:
            return 1
obj = Solution()
print(obj.countOperations(num1 = 2, num2 = 3))

3


In [ ]:
class Solution:
    def countOperations(self, num1: int, num2: int) -> int:
        op = 0
        while True:
            if num1 > num2:
                num1 -= num2
                op += 1
            elif num2 > num1:
                num2 -= num1
                op += 1
            else:
                num1 = 0
                op += 1
            if num1 == 0 or num2 == 0:
                break
        return op
            


obj = Solution()
print(obj.countOperations(num1 = 79, num2 = 68))

14


## **118. Pascal's Triangle**

In [ ]:
class Solution:
    def generate(self, numRows: int):
        if numRows == 1:
            return [[1]]
        lst = [[1]]
        i = 0
        for c in range(numRows-1):
            lst1 = [1]
            for j in range(len(lst[i])-1):
                lst1.append(lst[i][j] + lst[i][j+1])
            lst.append(lst1 + [1])
            i += 1
        return lst
obj = Solution()
print(obj.generate(30))

[[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1], [1, 5, 10, 10, 5, 1], [1, 6, 15, 20, 15, 6, 1], [1, 7, 21, 35, 35, 21, 7, 1], [1, 8, 28, 56, 70, 56, 28, 8, 1], [1, 9, 36, 84, 126, 126, 84, 36, 9, 1], [1, 10, 45, 120, 210, 252, 210, 120, 45, 10, 1], [1, 11, 55, 165, 330, 462, 462, 330, 165, 55, 11, 1], [1, 12, 66, 220, 495, 792, 924, 792, 495, 220, 66, 12, 1], [1, 13, 78, 286, 715, 1287, 1716, 1716, 1287, 715, 286, 78, 13, 1], [1, 14, 91, 364, 1001, 2002, 3003, 3432, 3003, 2002, 1001, 364, 91, 14, 1], [1, 15, 105, 455, 1365, 3003, 5005, 6435, 6435, 5005, 3003, 1365, 455, 105, 15, 1], [1, 16, 120, 560, 1820, 4368, 8008, 11440, 12870, 11440, 8008, 4368, 1820, 560, 120, 16, 1], [1, 17, 136, 680, 2380, 6188, 12376, 19448, 24310, 24310, 19448, 12376, 6188, 2380, 680, 136, 17, 1], [1, 18, 153, 816, 3060, 8568, 18564, 31824, 43758, 48620, 43758, 31824, 18564, 8568, 3060, 816, 153, 18, 1], [1, 19, 171, 969, 3876, 11628, 27132, 50388, 75582, 92378, 92378, 75582, 50388, 27132, 11628, 3876

In [ ]:
class Solution:
    def generate(self, numRows: int):
        if numRows == 1:
            return [[1]]
        elif numRows == 2:
            return [[1],[1,1]]
        lst = [[1],[1,1]]
        i = 1
        for c in range(numRows-2):
            lst1 = [1]
            for j in range(len(lst[i])-1):
                lst1.append(lst[i][j] + lst[i][j+1])
            lst.append(lst1 + [1])
            i += 1
        return lst
obj = Solution()
print(obj.generate(5))

[[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1]]


## **551. Student Attendance Record I**

In [ ]:
%%time
class Solution:
    def checkRecord(self, s: str) -> bool:
        if 'LLL' in s:
            return False
        if 'A' in s:
            i = s.index('A')
            s = s[:i] + s[i+1:]
        if 'A' in s:
            return False
        return True
obj = Solution()
print(obj.checkRecord(s = "PPALLP"))

True
CPU times: user 142 µs, sys: 0 ns, total: 142 µs
Wall time: 148 µs


In [ ]:
%%time
class Solution:
    def checkRecord(self, s: str) -> bool:
        if s.count('A') < 2:
            if 'LLL' in s:
                return False
            else:
                return True
        else:
            return False
obj = Solution()
print(obj.checkRecord(s = "PPALLP"))

True
CPU times: user 159 µs, sys: 0 ns, total: 159 µs
Wall time: 125 µs


# **Daily Coding Challenge**

## ***Feb 6:* 80. Remove Duplicates from Sorted Array II**

In [ ]:
class Solution:
    def removeDuplicates(self, nums) -> int:
        l = len(nums)
        prev = nums[0]
        count = 0
        d_count = 0
        for i in range(l-1):
            if d_count >= 2:
                nums.remove(nums[i])
                nums.append(0)
            else:
                count += 1
            if nums[i+1] != nums[i]:
                d_count = 0
            else:
                d_count +=1
        print(nums)
        return 5

In [ ]:
class Solution:
    def removeDuplicates(self, nums) -> int:
        if len(nums) < 3: 
            return len(nums)
        pos = 1
        for i in range(1, len(nums)-1):
            if nums[i-1] != nums[i+1]:
                nums[pos] = nums[i]
                pos += 1
        nums[pos] = nums[-1]
        print(nums)
        return pos + 1

obj = Solution()
print(obj.removeDuplicates(nums = [0,0,1,1,1,1,2,3,3]))

[0, 0, 1, 1, 2, 3, 3, 3, 3]
7


## ***Feb 7:* 389. Find the Difference**

In [ ]:
class Solution:

    def findTheDifference(self, s: str, t: str) -> str:

        sum1 = 0

        sum2 = 0

        for i in s:

            sum1 += ord(i)

        for i in t:

            sum2 += ord(i)

        return chr(sum2 - sum1)

obj = Solution()
print(obj.findTheDifference(s = "abcd", t = "abcde"))

e


## ***Feb 9:* 258. Add Digits**

In [ ]:
def make_list(n):
    lst = list(str(n))
    lst = list(map(int, lst))
    return 
make_list(399)

<class 'int'>


In [ ]:
class Solution:
    def addDigits(self, num: int) -> int:
        def make_list(n):
            lst = list(str(n))
            lst = list(map(int, lst))
            return lst
        while num>9:
            lst = make_list(num)
            val = 0
            for i in lst:
                val += i
            num = val
        return num
obj = Solution()
print(obj.addDigits(398))

2


## ***Feb 9:* 532. K-diff Pairs in an Array**

In [ ]:
class Solution:
    def findPairs(self, nums, k: int) -> int:
        # return the number of unique pairs
        answ=0
        # create a dictionary: d[x]==nums.count(x)
        d={}
        for x in nums:
            if x in d: d[x]+=1
            else:      d[x]=1
        
        if k: # k>0
            answ=sum(x+k in d for x in d.keys())
        else: # k==0
            answ=sum(k>1 for k in d.values())
                
        return answ
obj = Solution()
print(obj.findPairs(nums = [3,1,4,1,5], k = 2))

2


## ***Feb 10:* 560. Subarray Sum Equals K**

In [ ]:
class Solution:
    def subarraySum(self, nums, k: int) -> int:
        d ={0:1}
    
        sums = 0

        count= 0

        for i in range(len(nums)):
            sums+= nums[i]

            if sums-k in d:
                count+= d[sums-k]

            if sums in d:
                d[sums]+=1

            else:
                d[sums] = 1

        return count
obj = Solution()
print(obj.subarraySum(nums = [1,2,3], k = 3))

2


## ***Feb 11:* 567. Permutation in String**

In [ ]:
import itertools
s = 'ABC'
nums = list(s)
# nums = ['ABC']
permutations = list(itertools.permutations(nums))
lst = [''.join(permutation) for permutation in permutations]
print(lst)

['ABC', 'ACB', 'BAC', 'BCA', 'CAB', 'CBA']


In [ ]:
from collections import Counter
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        # Edge Condition
        if len(s1) > len(s2):
            return False
        n1 = len(s1)
        
		# Hashmap for s1
        c1 = Counter(s1)
		
        for i in range(len(s2)-n1+1):
            if Counter(s2[i:i+n1]) == c1:
                       return True
        return False
obj = Solution()
print(obj.checkInclusion(s1 = "ab", s2 = "eidbaooo"))

True


## ***Feb 12:* 127. Word Ladder**

In [ ]:
from collections import *
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList) -> int:
        if endWord not in wordList or not endWord or not beginWord or not wordList:
            return 0
        L = len(beginWord)
        all_combo_dict = defaultdict(list)
        for word in wordList:
            for i in range(L):
                all_combo_dict[word[:i] + "*" + word[i+1:]].append(word) 
        queue = deque([(beginWord, 1)])
        visited = set()
        visited.add(beginWord)
        while queue:
            current_word, level = queue.popleft()
            for i in range(L):
                intermediate_word = current_word[:i] + "*" + current_word[i+1:]
                for word in all_combo_dict[intermediate_word]:
                    if word == endWord:
                        return level + 1
                    if word not in visited:
                        visited.add(word)
                        queue.append((word, level + 1))
        return 0
obj = Solution()
print(obj.ladderLength(beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log","cog"]))

5


## ***Feb 13:* 78. Subsets**

In [ ]:
class Solution:
    def subsets(self, nums):
        nums.sort()
        lst = [[]]
        for n in nums:
            lst += [i + [n] for i in lst]
        return lst
obj = Solution()
print(obj.subsets(nums = [1,2,3]))

[[], [1], [2], [1, 2], [3], [1, 3], [2, 3], [1, 2, 3]]


## ***Feb 14:* 104. Maximum Depth of Binary Tree**

In [ ]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def maxDepth(self, root) -> int:
        def dfs(root, depth):
            if not root: return depth
            return max(dfs(root.left, depth + 1), dfs(root.right, depth + 1))

        return dfs(root, 0)
obj = Solution()

ob2 = TreeNode(9, None, None)
ob4 = TreeNode(7, None, None)
ob5 = TreeNode(15, None, None)
ob3 = TreeNode(20, ob5, ob4)
ob1 = TreeNode(3, ob2, ob3)

print(obj.maxDepth(ob1))

3


## ***Feb 15:*136. Single Number**

In [ ]:
class Solution:
    def singleNumber(self, nums) -> int:
        return (2 * sum(set(nums))) - sum(nums)
obj = Solution()
print(obj.singleNumber(nums = [4,1,2,1,2]))

4


## ***Feb 17:* 39. Combination Sum**

In [ ]:
class Solution:
    def combinationSum(self, candidates, target: int):
        idx_d = {val: idx for idx, val in enumerate(candidates)} # create {num:idx}
        n = len(candidates)
        dp = [[] for _ in range(target+1)]
        for i in range(1, target+1):                         # from 1 to target, O(M)
            for j in range(i):                               # O(M): for all previous calculated numbers
                for comb in dp[j]:                           # O(M) worst: check each of their combinations
                    start_idx = idx_d[comb[-1]]              # use start_idx to avoid repeats
                    for val in candidates[start_idx:]:       # O(N): try all candidates
                        if val + j == i: dp[i].append(comb + [val]) 
            for candidate in candidates:                     # O(N): directly from candidates not from previous result 
                if candidate == i: dp[i].append([candidate])
        return dp[-1]
obj = Solution()
print(obj.combinationSum(candidates = [2,3,6,7], target = 7))

[[2, 2, 3], [7]]


## ***Feb 18:* 402. Remove K Digits**

In [ ]:
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []

        for n in num:
            while stack and k > 0 and stack[-1] > n:
                k -= 1
                stack.pop()
            stack.append(n)

        while stack and k > 0:
            stack.pop()
            k -= 1

        if not stack:
            return "0"

        return str(int("".join(stack)))
obj = Solution()
print(obj.removeKdigits(num = "1432219", k = 3))

1219


## ***Feb 19:* 1675. Minimize Deviation in Array**

In [ ]:
MAX = 10**20
from heapq import *
class Solution:
    def minimumDeviation(self, nums) -> int:
        heap = []
        answer = MAX
        array_length = len(nums)
        
        for num in nums:
            if num % 2 == 0:
                heappush(heap, -num)
            else:
                heappush(heap, -num*2)
        
        _min = -max(heap)

        while True:
            _max = -heappop(heap)
            answer = min(answer, _max-_min)
            
            if _max % 2:
                break

            heappush(heap, -_max//2)
            _min = min(_min, _max//2)

        return answer
obj = Solution()
print(obj.minimumDeviation(nums = [1,2,3,4]))

1


## ***Feb 20:* 1288. Remove Covered Intervals**

In [ ]:
class Solution:
    def removeCoveredIntervals(self, intervals) -> int:
        intervals = sorted(intervals, key = lambda x : (x[0], -x[1]))
        count = len(intervals)
        l = intervals[0][0]
        h = intervals[0][1]
        for i in range(1, len(intervals)):
            if intervals[i][0] < l and intervals[i][1] > h:
                l = intervals[i][0]
                h = intervals[i][1]
                count -= 1
                print(intervals[i], end = ' ')
            elif (intervals[i][0] and intervals[i][1]) in range(l, h+1):
                count -= 1
            elif intervals[i][1] > h:
                h = intervals[i][1]
            elif intervals[i][0] < l:
                l = intervals[i][0]
        return count
                

        
obj = Solution()
intervals = [[1,4],[3,6],[2,8]]
intervals = [[1,4],[2,3]]
intervals = [[34335,39239],[15875,91969],[29673,66453],[53548,69161],[40618,93111]]
# intervals.sort(key=lambda x: [x[0], -x[1]])
# print(intervals)
print(obj.removeCoveredIntervals(intervals))

[40618, 93111]
2


## ***Feb 22:* 171. Excel Sheet Column Number**

## ***Feb 23:* 133. Clone Graph**

## ***Feb 24:* Sort List**

# ***Feb 25:* 165. Compare Version Numbers**

## ***Feb 26:* 847. Shortest Path Visiting All Nodes**

In [9]:
class Solution:
    def shortestPathLength(self, graph) -> int:
        length = len(graph)

        result = 0

        visited_node = []

        queue = []  

        for i in range(length):
          visited_node.append(set([1<<i]))
          queue.append([i,1<<i])

        while queue:

          result = result + 1

          new_queue = []

          for node, value in queue:

            for neigbour_node in graph[node]:

              mid_node = (1<<neigbour_node)|value

              if mid_node not in visited_node[neigbour_node]:

                if mid_node+1 == 1<<length:

                  return result

                visited_node[neigbour_node].add(mid_node)

                new_queue.append([neigbour_node, mid_node])

          queue = new_queue

        return 0
obj = Solution()
print(obj.shortestPathLength(graph = [[1],[0,2,4],[1,3,4],[2],[1,2]]))

4


# **Biweekly Contest 72**

## **2176. Count Equal and Divisible Pairs in an Array**

In [ ]:
class Solution:
    def countPairs(self, nums, k: int) -> int:
        if list(set(nums)) == nums:
            return 0
        l = len(nums)
        count = 0
        for i in range(l):
            for j in range(i+1 ,l):
                if nums[i] == nums[j]:
                    if (i * j) % k == 0:
                        count += 1
                        print(i,j)
        return count

obj = Solution()
print(obj.countPairs(nums = [3,1,2,2,2,1,3], k = 2))

0 6
2 3
2 4
3 4
4


## **2177. Find Three Consecutive Integers That Sum to a Given Number**

In [ ]:
n = 100
val = n // 3
print([val-1 ,val , val+1], sum([val-1 ,val , val+1]))

[32, 33, 34] 99


## **2178. Maximum Split of Positive Even Integers**

In [ ]:
class Solution:
    def maximumEvenSplit(self, finalSum: int):
        if finalSum % 2 != 0:
            return []
        x = 2
        ret = []
        while finalSum >= x * 2 + 2:
            ret.append(x)
            finalSum -= x
            x += 2
            print(x, finalSum, x * 2 + 2)
            ret.append(finalSum)
        return ret
obj = Solution()
print(obj.maximumEvenSplit(finalSum = 28))

4 26 10
6 22 14
8 16 18
[2, 26, 4, 22, 6, 16]


# **Weekly Contest 281**

## **6012. Count Integers With Even Digit Sum**

In [ ]:
class Solution:
    def countEven(self, num: int) -> int:
        def is_even(n):
            val = str(n)
            ret = 0
            for i in val:
                ret += int(i)
            #     print(i, end = ' ')
            # print(ret, end = ' ')
            if ret % 2 == 0:
                return True
            else:
                return False
        
        count = 0
        for i in range(1, num+1):
            if is_even(i):
                print(i)
                count += 1
        return count
obj = Solution()
print(obj.countEven(30))

2
4
6
8
11
13
15
17
19
20
22
24
26
28
14


## **6013. Merge Nodes in Between Zeros**

In [ ]:
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def print_Node(self):
        lst = []
        while self:
          lst.append(self.val)
          self = self.next
        return lst
class Solution:
    def mergeNodes(self, head):
        curr = head
        ret = ListNode(0)
        ret1 = ret
        array = []
        curr = curr.next
        while curr:
            if curr.val != 0:
                array.append(curr.val)
                curr = curr.next
            elif curr.val == 0:
                cur = ListNode(sum(array))
                ret.next = cur
                ret = ret.next
                curr = curr.next
                array = []
        return ret1.next
o1 = ListNode()
o2 = ListNode()
o3 = ListNode()
o4 = ListNode()
o5 = ListNode()
o6 = ListNode()
o7 = ListNode()
o8 = ListNode()
o1.val = 0
o2.val = 1
o3.val = 0
o4.val = 3
o5.val = 0
o6.val = 2
o7.val = 2
o8.val = 0
head = o1
o1.next = o2
o2.next = o3
o3.next = o4
o4.next = o5
o5.next = o6
o6.next = o7
o7.next = o8
o8.next = None

obj = Solution()
head1 = obj.mergeNodes(head)
head1.print_Node()
# print(obj.mergeNodes(head1))


[1, 3, 4]